8_UI/Leaderboard/ViewControllers/LeaderboardView.cs

```diff
@@ -9,36 +9,47 @@ internal class LeaderboardView : BSMLAutomaticViewController {
         #region Components
 
         [UIValue("voting-panel"), UsedImplicitly]
-        private VotingPanel _votingPanel = ReeUIComponentV2.InstantiateOnSceneRoot<VotingPanel>(false);
+        private VotingPanel _votingPanel;
 
         [UIValue("beat-leader-info"), UsedImplicitly]
-        private BeatLeaderInfo _beatLeaderInfo = ReeUIComponentV2.InstantiateOnSceneRoot<BeatLeaderInfo>(false);
+        private BeatLeaderInfo _beatLeaderInfo;
 
         [UIValue("leaderboard-settings"), UsedImplicitly]
-        private LeaderboardSettings _leaderboardSettings = ReeUIComponentV2.InstantiateOnSceneRoot<LeaderboardSettings>(false);
+        private LeaderboardSettings _leaderboardSettings;
 
         [UIValue("score-info-panel"), UsedImplicitly]
-        private ScoreInfoPanel _scoreInfoPanel = ReeUIComponentV2.InstantiateOnSceneRoot<ScoreInfoPanel>(false);
+        private ScoreInfoPanel _scoreInfoPanel;
 
         [UIValue("scores-table"), UsedImplicitly]
-        private ScoresTable _scoresTable = ReeUIComponentV2.InstantiateOnSceneRoot<ScoresTable>();
+        private ScoresTable _scoresTable;
 
         [UIValue("voting-button"), UsedImplicitly]
-        private VotingButton _votingButton = ReeUIComponentV2.InstantiateOnSceneRoot<VotingButton>(false);
+        private VotingButton _votingButton;
 
         [UIValue("pagination"), UsedImplicitly]
-        private Pagination _pagination = ReeUIComponentV2.InstantiateOnSceneRoot<Pagination>(false);
+        private Pagination _pagination;
 
         [UIValue("scope-selector"), UsedImplicitly]
-        private ScopeSelector _scopeSelector = ReeUIComponentV2.InstantiateOnSceneRoot<ScopeSelector>(false);
+        private ScopeSelector _scopeSelector;
 
         [UIValue("context-selector"), UsedImplicitly]
-        private ContextSelector _contextSelector = ReeUIComponentV2.InstantiateOnSceneRoot<ContextSelector>(false);
+        private ContextSelector _contextSelector;
 
         [UIValue("empty-board-message"), UsedImplicitly]
-        private EmptyBoardMessage _emptyBoardMessage = ReeUIComponentV2.InstantiateOnSceneRoot<EmptyBoardMessage>();
+        private EmptyBoardMessage _emptyBoardMessage;
 
         private void Awake() {
```

Comment:
Components were instantiated and parsed before Awake() on purpose: to pre-parse and initialize layout while the game is still loading, so player won't notice.

Awake on the other hand is called only when leaderboard is shown to player, and huge lag spike there is very noticeable and annoying (it takes several seconds)

Reply:
There must be another way of achieving the same result without the lag spike and breaking the game outside of production.

The problematic line is this one:

https://github.com/BeatLeader/beatleader-mod/blob/e666219572f8ac3e198eeb4697552a341f347102/8_UI/ReeUIComponentV2.cs#L49

Unity throws this on a debug build:

> UnityException: get_sceneCount is not allowed to be called from a MonoBehaviour constructor (or instance field initializer), call it in Awake or Start instead.

There is probably another way of getting the last loaded scene, but due to a mix of debug limitations/static context limitations, I can't wrap my head around it.

Reply:
Check the latest commit, please. I moved pre-parse logic to a separate monoBehavior. Should be enough to fix your issues

Reply:
Hey, even if this is enough, reopen the pull only with the Emulator fix. Sorry that I closed your pull accidentally btw

Reply:
That is indeed enough.